File: java/springboot/appfolder/src/main/java/juchoi/template/appfolder/transport/tcp/server/serializer/BinaryWriter.java
package juchoi.template.appfolder.transport.tcp.server.serializer;

import juchoi.template.appfolder.utils.ByteUtils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteOrder;

// Packet format: [opcode 1B][length 2B big-endian][data length-B]
public class BinaryWriter {

    public static byte[] build(byte opcode, byte[] data) throws IOException {
        ByteArrayOutputStream buf = new ByteArrayOutputStream(3 + data.length);
        buf.write(opcode);
        buf.write(ByteUtils.fromShort((short) data.length, ByteOrder.BIG_ENDIAN));
        buf.write(data);
        return buf.toByteArray();
    }
}

File: c++/drogon/internal/transport/tcp/server/serializer/binary_writer.cpp
#include "internal/transport/tcp/server/serializer/binary_writer.h"

namespace app {

std::vector<uint8_t> BinaryWriter::build(uint8_t opcode,
                                         const std::string& body) {
    uint32_t len = static_cast<uint32_t>(body.size());
    std::vector<uint8_t> buf;
    buf.reserve(5 + len);

    buf.push_back(opcode);
    buf.push_back(static_cast<uint8_t>((len >> 24) & 0xFF));
    buf.push_back(static_cast<uint8_t>((len >> 16) & 0xFF));
    buf.push_back(static_cast<uint8_t>((len >> 8)  & 0xFF));
    buf.push_back(static_cast<uint8_t>((len)       & 0xFF));
    buf.insert(buf.end(), body.begin(), body.end());

    return buf;
}

} // namespace app
